src/main/java/com/xjeffrose/xio/http/RequestBuffer.java

```diff
@@ -0,0 +1,92 @@
+package com.xjeffrose.xio.http;
+
+import com.xjeffrose.xio.client.ClientState;
+import com.xjeffrose.xio.client.DefaultChannelInitializer;
+import com.xjeffrose.xio.client.XioClient;
+import com.xjeffrose.xio.client.XioClientBootstrap;
+import com.xjeffrose.xio.server.Route;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.Http2DataFrame;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.AttributeKey;
+import io.netty.util.ReferenceCountUtil;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import java.util.function.BiFunction;
+
+@Slf4j
+public class RequestBuffer extends ChannelDuplexHandler {
+
+  public static class WriteReady {
+    public static final WriteReady INSTANCE = new WriteReady();
+  }
+
+  public static class ObjectAndPromise {
+    public final Object object;
+    public final ChannelPromise promise;
+
+    ObjectAndPromise(Object object, ChannelPromise promise) {
+      this.object = object;
+      this.promise = promise;
+    }
+
+    public ChannelFuture apply(BiFunction<Object, ChannelPromise, ChannelFuture> write) {
+      return write.apply(object, promise);
+    }
+  }
+
+  List<ObjectAndPromise> writeBuffer = new ArrayList<>();
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof WriteReady)) {
+      ctx.fireUserEventTriggered(evt);
+      return;
+    }
+
+    // drain buffer
+    while (writeBuffer.size() > 0) {
+      ObjectAndPromise pair = writeBuffer.remove(0);
```

Comment:
`O(n)` for `ArrayList`. Use a `LinkedList` instead?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html

> The size, isEmpty, get, set, iterator, and listIterator operations run in constant time. The add operation runs in amortized constant time, that is, adding n elements requires O(n) time. All of the other operations run in linear time (roughly speaking). The constant factor is low compared to that for the LinkedList implementation.

Are you thinking of [ArrayDeque](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayDeque.html) ?

Reply:
`remove(0)` is linear-time per the docs you quote. Suggestion still stands?

Reply:
I meant should we use the ArrayDeque, not LinkedList.